File: src/ios/models/Geofence.m
/**
 * Geofence Superclass
 *
 * @copyright Center for Health Enhancement Systems Studies
 */
#import "Geofence.h"

@implementation Geofence
/**
 * Identifier
 *
 * @var NSString*
 */
NSString* __id = @"";

/**
 * State of geofence
 *
 * @var NSString*
 */
NSString* _state = @"";

/**
 * Initialize object
 *
 * @param new_id		Identifier for object
 *
 * @return Geofence
 */
- (id) init: (NSString*) new_id {
	self = [super init];
	self.__id = new_id;

	return self;
}

/**
 * Return state value
 *
 * @return NSString*
 */
- (NSString*) getState {
	return self._state;
}

/**
 * Check if point is within the geofence
 *
 * @param point			Point to check
 *
 * @return BOOL
 */
- (BOOL) isPointWithinGeofence: (CLLocation*) point {
	[NSException raise:@"`isPointWithinGeofence` must be overwritten in subclass" format:@"isPointWithinGeofence` must be overwritten in subclass"];
	return NO;
}

/**
 * Set state of the geofence
 *
 * @param new_state		State to set
 *
 * @return void
 */
- (void) setState: (NSString*) new_state {
	self._state = new_state;
}

@end

File: src/android/models/Geofence.java
/**
 * Geofence super class
 *
 * @copyright Center for Health Enhancement Systems Studies
 */
package edu.wisc.chess.plugins.chesslocationmonitoring.models;

import android.location.Location;

public class Geofence implements GeofenceInterface {
	/**
	 * Identifier for geofence
	 *
	 * @var String
	 */
	protected String id = "";

	/**
	 * State of the geofence
	 *
	 * @var String
	 */
	protected String state = "";

	/**
	 * Construct geofence
	 *
	 * @param id Identifier for geofence
	 *
	 * @return void
	 */
	public Geofence(String id) {
		this.id = id;
	}

	/**
	 * Return geofence state
	 *
	 * @return String
	 */
	public String getState() {
		return this.state;
	}

	/**
	 * Check if point is within the geofence
	 *
	 * @param gps_point Point to check
	 *
	 * @return boolean
	 */
	public boolean isPointWithinGeofence(Location gps_point) throws Exception {
		throw new Exception("`isPointWithinGeofence` must be overwritten in subclass");
	}

	/**
	 * Set geofence state
	 *
	 * @param state State to set
	 *
	 * @return void
	 */
	public void setState(String state) {
		this.state = state;
	}
}
